lib/internal/streams/pipeline.js

```diff
@@ -321,4 +321,40 @@ function pipelineImpl(streams, callback, opts) {
   return ret;
 }
 
+function pipe (src, dst) {
+  const state = src._readableState;
+
+  if (typeof src.read !== 'function' && state) {
+    src.pipe(dst);
+    return;
+  }
+
+  src
+    .on('end', end)
+    .on('readable', pump);
+  dst
+    .on('drain', pump);
+
+  function end () {
+    dst.end();
+  }
+
+  function pump () {
+    if (dst.writableNeedDrain) {
+      return;
+    }
+
+    while (true) {
+      state.highwaterMark = dst.writableHighwaterMark || dst.highwaterMark || null;
```

Comment:
I think there is a bug in this logic with mixed binary/object streams. If the writable is in objectMode they will have 16 (or similar) as a highwaterMark. However if the readable is binary, it will be lowered significantly, slowing everything to a crawl.